File: services/chess_service/src/main/java/net/chess_platform/chess_service/chess/piece/Queen.java
package net.chess_platform.chess_service.chess.piece;

import java.util.ArrayList;
import java.util.List;

import net.chess_platform.chess_service.chess.Chessboard;
import net.chess_platform.chess_service.chess.move.Move;
import net.chess_platform.chess_service.chess.piece.behavior.BishopBehavior;
import net.chess_platform.chess_service.chess.piece.behavior.PieceBehavior;
import net.chess_platform.chess_service.chess.piece.behavior.RookBehavior;

public class Queen extends AbstractPiece {

    private PieceBehavior rookMoves = new RookBehavior();

    private PieceBehavior bishopMoves = new BishopBehavior();

    public Queen(Color color) {
        super(color, Type.QUEEN);
    }

    @Override
    public List<Move> getMoves(Chessboard board, int row, int col) {
        var rookMoves_ = rookMoves.getMoves(board, getType(), getColor(), row, col);
        var bishopMoves_ = bishopMoves.getMoves(board, getType(), getColor(), row, col);

        var moveList = new ArrayList<Move>(rookMoves_);
        moveList.addAll(bishopMoves_);

        return moveList;
    }

    @Override
    public String toString() {
        return getColor() == Color.WHITE ? Character.toString(9813) : Character.toString(9819);
    }
}

File: app/src/chess/pieces/Queen.ts
import { Chessboard } from "../Chessboard";
import { Move } from "../moves/Move";
import { AbstractPiece, Color } from "./AbstractPiece";
import { BishopBehavior } from "./behavior/BishopBehavior";
import { RookBehavior } from "./behavior/RookBehavior";

export class Queen extends AbstractPiece {

    private rookBehavior = new RookBehavior();

    private bishopBehavior = new BishopBehavior();

    constructor(color: Color, moveCount?: number) {
        super(color, "QUEEN", moveCount);
    }

    getLegalMovesImpl(board: Chessboard, row: number, col: number): Move[] {
        return [
            ...this.rookBehavior.getMoves(board, this.getType(), this.getColor(), row, col),
            ...this.bishopBehavior.getMoves(board, this.getType(), this.getColor(), row, col)
        ];
    }

    override toString(): string {
        return this.getColor() == "WHITE" ? String.fromCodePoint(9813) : String.fromCodePoint(9819);
    }
}